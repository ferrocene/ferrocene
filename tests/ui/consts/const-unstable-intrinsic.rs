//! Ensure that unstable intrinsics can actually not be called,
//! neither within a crate nor cross-crate.
//@ aux-build:unstable_intrinsic.rs
#![feature(staged_api, rustc_attrs, intrinsics)]
#![stable(since="1.0.0", feature = "stable")]
#![feature(local)]

extern crate unstable_intrinsic;

fn main() {
    const_main();
}

const fn const_main() {
    let x = 42;
    unsafe {
        unstable_intrinsic::old_way::size_of_val(&x);
        //~^ERROR: unstable library feature 'unstable'
        //~|ERROR: cannot call non-const intrinsic
        unstable_intrinsic::old_way::min_align_of_val(&x);
        //~^ERROR: unstable library feature 'unstable'
        //~|ERROR: not yet stable as a const intrinsic
        unstable_intrinsic::new_way::size_of_val(&x);
        //~^ERROR: unstable library feature 'unstable'
        //~|ERROR: cannot be (indirectly) exposed to stable
        unstable_intrinsic::new_way::min_align_of_val(&x);
        //~^ERROR: unstable library feature 'unstable'
        //~|ERROR: not yet stable as a const intrinsic

        old_way::size_of_val(&x);
        //~^ERROR: cannot call non-const intrinsic
        old_way::min_align_of_val(&x);
        //~^ERROR: cannot use `#[feature(local)]`
        new_way::size_of_val(&x);
        //~^ERROR: cannot be (indirectly) exposed to stable
        new_way::min_align_of_val(&x);
        //~^ERROR: cannot use `#[feature(local)]`
    }
}

#[stable(since="1.0.0", feature = "stable")]
pub mod old_way {
    extern "rust-intrinsic" {
        #[unstable(feature = "local", issue = "42")]
        pub fn size_of_val<T>(x: *const T) -> usize;

        #[unstable(feature = "local", issue = "42")]
        #[rustc_const_unstable(feature = "local", issue = "42")]
        pub fn min_align_of_val<T>(x: *const T) -> usize;
    }
}

#[stable(since="1.0.0", feature = "stable")]
pub mod new_way {
    #[unstable(feature = "local", issue = "42")]
    #[rustc_intrinsic]
    pub const unsafe fn size_of_val<T>(x: *const T) -> usize { 42 }

    #[unstable(feature = "local", issue = "42")]
    #[rustc_const_unstable(feature = "local", issue = "42")]
    #[rustc_intrinsic]
    pub const unsafe fn min_align_of_val<T>(x: *const T) -> usize { 42 }
}

#[stable(feature = "rust1", since = "1.0.0")]
#[rustc_const_stable(feature = "const_intrinsic_copy", since = "1.63.0")]
#[inline]
pub const unsafe fn copy<T>(src: *const T, dst: *mut T, count: usize) {
    // Const stability attributes are not inherited from parent items.
    extern "rust-intrinsic" {
        fn copy<T>(src: *const T, dst: *mut T, count: usize);
    }

    unsafe { copy(src, dst, count) }
    //~^ ERROR cannot call non-const intrinsic
}
