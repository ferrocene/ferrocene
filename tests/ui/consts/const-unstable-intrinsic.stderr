error[E0658]: use of unstable library feature 'unstable'
  --> $DIR/const-unstable-intrinsic.rs:17:9
   |
LL |         unstable_intrinsic::old_way::size_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #42 <https://github.com/rust-lang/rust/issues/42> for more information
   = help: add `#![feature(unstable)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'unstable'
  --> $DIR/const-unstable-intrinsic.rs:20:9
   |
LL |         unstable_intrinsic::old_way::min_align_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #42 <https://github.com/rust-lang/rust/issues/42> for more information
   = help: add `#![feature(unstable)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'unstable'
  --> $DIR/const-unstable-intrinsic.rs:23:9
   |
LL |         unstable_intrinsic::new_way::size_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #42 <https://github.com/rust-lang/rust/issues/42> for more information
   = help: add `#![feature(unstable)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'unstable'
  --> $DIR/const-unstable-intrinsic.rs:26:9
   |
LL |         unstable_intrinsic::new_way::min_align_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #42 <https://github.com/rust-lang/rust/issues/42> for more information
   = help: add `#![feature(unstable)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date

error: cannot call non-const intrinsic `size_of_val` in constant functions
  --> $DIR/const-unstable-intrinsic.rs:17:9
   |
LL |         unstable_intrinsic::old_way::size_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `min_align_of_val` is not yet stable as a const intrinsic
  --> $DIR/const-unstable-intrinsic.rs:20:9
   |
LL |         unstable_intrinsic::old_way::min_align_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: add `#![feature(unstable)]` to the crate attributes to enable

error: intrinsic `unstable_intrinsic::new_way::size_of_val` cannot be (indirectly) exposed to stable
  --> $DIR/const-unstable-intrinsic.rs:23:9
   |
LL |         unstable_intrinsic::new_way::size_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: mark the caller as `#[rustc_const_unstable]`, or mark the intrinsic `#[rustc_const_stable_indirect]` (but this requires team approval)

error: `min_align_of_val` is not yet stable as a const intrinsic
  --> $DIR/const-unstable-intrinsic.rs:26:9
   |
LL |         unstable_intrinsic::new_way::min_align_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: add `#![feature(unstable)]` to the crate attributes to enable

error: cannot call non-const intrinsic `size_of_val` in constant functions
  --> $DIR/const-unstable-intrinsic.rs:30:9
   |
LL |         old_way::size_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^

error: const function that might be (indirectly) exposed to stable cannot use `#[feature(local)]`
  --> $DIR/const-unstable-intrinsic.rs:32:9
   |
LL |         old_way::min_align_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: if the function is not (yet) meant to be exposed to stable, add `#[rustc_const_unstable]` (this is what you probably want to do)
   |
LL + #[rustc_const_unstable(feature = "...", issue = "...")]
LL | const fn const_main() {
   |
help: otherwise, as a last resort `#[rustc_allow_const_fn_unstable]` can be used to bypass stability checks (this requires team approval)
   |
LL + #[rustc_allow_const_fn_unstable(local)]
LL | const fn const_main() {
   |

error: intrinsic `new_way::size_of_val` cannot be (indirectly) exposed to stable
  --> $DIR/const-unstable-intrinsic.rs:34:9
   |
LL |         new_way::size_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: mark the caller as `#[rustc_const_unstable]`, or mark the intrinsic `#[rustc_const_stable_indirect]` (but this requires team approval)

error: const function that might be (indirectly) exposed to stable cannot use `#[feature(local)]`
  --> $DIR/const-unstable-intrinsic.rs:36:9
   |
LL |         new_way::min_align_of_val(&x);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
help: if the function is not (yet) meant to be exposed to stable, add `#[rustc_const_unstable]` (this is what you probably want to do)
   |
LL + #[rustc_const_unstable(feature = "...", issue = "...")]
LL | const fn const_main() {
   |
help: otherwise, as a last resort `#[rustc_allow_const_fn_unstable]` can be used to bypass stability checks (this requires team approval)
   |
LL + #[rustc_allow_const_fn_unstable(local)]
LL | const fn const_main() {
   |

error: cannot call non-const intrinsic `copy` in constant functions
  --> $DIR/const-unstable-intrinsic.rs:74:14
   |
LL |     unsafe { copy(src, dst, count) }
   |              ^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 13 previous errors

For more information about this error, try `rustc --explain E0658`.
